### PR TITLE
howto: Add VMCache introduction and guide

### PR DESCRIPTION
**Pull request by teawater**
Add "What Is VMCache and How To Enable It" to howto.

Fixes: #392

Signed-off-by: Hui Zhu <teawater@hyper.sh>

### CODE REVIEW

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# What Is VMCache and How To Enable It
+
+### What is VMCache
+VMCache is a new function that creates VMs as caches before using it.
+It helps speed up new container creation.
+The function consists of a server and some clients communicating
+through Unix socket.  The protocol is GRPC in protocols/cache/cache.proto.
```
Comment (jodh-intel) [inline how-to/what-is-vm-cache-and-how-do-I-use-it.md]: - Nit: This should really be "gRPC" (throughout the document).
- Could you make `protocols/cache/cache.proto` a link to https://github.com/kata-containers/runtime/blob/master/protocols/cache/cache.proto? That won't be valid until https://github.com/kata-containers/runtime/pull/1166 lands of course but makes it easier for users to find.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+requestion from clients.
+Factory grpccache is the VMCache client.  It will request GRPC format
+VM and convert it back to a VM.  If VMCache function is enabled,
+kata-runtime will request VM from factory grpccache when it creates
+a new sandbox.
+
+### How to enable VMCache
+VMCache can be enabled by changing your Kata Containers config file (`/usr/share/defaults/kata-containers/configuration.toml`,
+overridden by `/etc/kata-containers/configuration.toml` if provided) such that:
+* `vm_cache_number` the number of caches of VMCache:
```
Comment (jodh-intel) [inline how-to/what-is-vm-cache-and-how-do-I-use-it.md]: s/the number/specifies the number/

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    * unspecified or == 0   --> VMCache is disabled
+    * `> 0`                   --> will be set to the specified number
+*  `vm_cache_endpoint` specifies the address of the Unix socket.
+
+Then you can create a VM templating for later usage by calling
```
Comment (jodh-intel) [inline how-to/what-is-vm-cache-and-how-do-I-use-it.md]: s/calling/calling:/

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+```
+$ sudo kata-runtime factory init
+```
+and purge it by ctrl-c it.
+
+Currently, VM cache still cannot work with VM templating and vsock.
```
Comment (jodh-intel) [inline how-to/what-is-vm-cache-and-how-do-I-use-it.md]: I'd make this into a new section. Something like:

> # Limitations
>
> Restrictions on the use of VM cache:
>
> - Cannot work with VM templating.
> - Cannot work with vsock.
> - Only supports the qemu hypervisor.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# What Is VMCache and How To Enable It
```
Comment (jodh-intel) [inline how-to/what-is-vm-cache-and-how-do-I-use-it.md]: Please could you add a table of contents at the top?

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+The function consists of a server and some clients communicating
+through Unix socket.  The protocol is GRPC in protocols/cache/cache.proto.
+The VMCache server will create some VMs and cache them by factory cache.
+It will convert the VM to GRPC format and transport it when gets
+requestion from clients.
+Factory grpccache is the VMCache client.  It will request GRPC format
+VM and convert it back to a VM.  If VMCache function is enabled,
+kata-runtime will request VM from factory grpccache when it creates
+a new sandbox.
+
```
Comment (jodh-intel) [inline how-to/what-is-vm-cache-and-how-do-I-use-it.md]: I think a common question is going to be "how is this different to VM templating?" so I would add a section to explain that and reference https://github.com/kata-containers/documentation/blob/master/how-to/what-is-vm-templating-and-how-do-I-use-it.md.

(Also, does https://github.com/kata-containers/documentation/blob/master/how-to/what-is-vm-templating-and-how-do-I-use-it.md need to reference this document?)

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md (hunk 1 above)
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > * Nit: This should really be "gRPC" (throughout the document).
> * Could you make `protocols/cache/cache.proto` a link to https://github.com/kata-containers/runtime/blob/master/protocols/cache/cache.proto? That won't be valid until [kata-containers/runtime#1166](https://github.com/kata-containers/runtime/pull/1166) lands of course but makes it easier for users to find.

Fixed.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md (hunk 2 above)
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > s/the number/specifies the number/

Fixed.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md (hunk 3 above)
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > s/calling/calling:/

Fixed.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md (hunk 4 above)
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > I'd make this into a new section. Something like:
> 
> > # Limitations
> > Restrictions on the use of VM cache:
> > 
> > * Cannot work with VM templating.
> > * Cannot work with vsock.
> > * Only supports the qemu hypervisor.

Added.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md (hunk 5 above)
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > Please could you add a table of contents at the top?

Added.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md (hunk 6 above)
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > I think a common question is going to be "how is this different to VM templating?" so I would add a section to explain that and reference https://github.com/kata-containers/documentation/blob/master/how-to/what-is-vm-templating-and-how-do-I-use-it.md.
> 
> (Also, does https://github.com/kata-containers/documentation/blob/master/how-to/what-is-vm-templating-and-how-do-I-use-it.md need to reference this document?)

Added.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+requestion from clients.  
+Factory grpccache is the VMCache client.  It will request gRPC format
+VM and convert it back to a VM.  If VMCache function is enabled,
+kata-runtime will request VM from factory grpccache when it creates
+a new sandbox.
+
+### How is this different to VM templating
+Both [VM templating](https://github.com/kata-containers/documentation/blob/master/how-to/what-is-vm-templating-and-how-do-I-use-it.md) and VMCache help speed up new container creation.  
+When VM templating enabled, new VMs are created by cloning from a pre-created template VM, and they will share the same initramfs, kernel and agent memory in readonly mode.  So it saves a lot of memory if there are many Kata Containers running on the same host.  
+VMCache is not vulnerable to share memory CVE because each VM doesn't share the memory.
```
Comment (sboeuf) [inline how-to/what-is-vm-cache-and-how-do-I-use-it.md]: Can you link the CVE here? Or at least the exact number.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
+# What Is VMCache and How To Enable It
+
+* [What is VMCache](#what-is-vmcache-and-how-to-enable-it)
+* [How is this different to VM templating](#how-is-this-different-to-vm-templating)
```
[the remark this answers is not in the document]
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > I would say: `How is this different from VM templating`

Fixed.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+* [Limitations](#limitations)
+
+### What is VMCache
+VMCache is a new function that creates VMs as caches before using it.
+It helps speed up new container creation.  
+The function consists of a server and some clients communicating
+through Unix socket.  The protocol is gRPC in [protocols/cache/cache.proto](https://github.com/kata-containers/runtime/blob/master/protocols/cache/cache.proto).  
+The VMCache server will create some VMs and cache them by factory cache.
+It will convert the VM to gRPC format and transport it when gets
+requestion from clients.
```
[the remark this answers is not in the document]
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > s/requestion/requested/

Fixed.

File: how-to/what-is-vm-cache-and-how-do-I-use-it.md (hunk 7 above)
Comment (teawater) [reply how-to/what-is-vm-cache-and-how-do-I-use-it.md]: > Can you link the CVE here? Or at least the exact number.

Fixed.

**Comment by gnawux** [comment]
nice document

**Pull request merged**